### PR TITLE
ScriptDebugger refactor, threading, profilers.

### PR DESCRIPTION
**Pull request by Faless**
Refactor ScriptDebugger (followup on #36244).
EngineDebugger is the new interface to access the debugger.
It tries to be as agnostic as possible on the data that various subsystems can expose.

It allows 2 types of interactions:

- Profilers:
  A subsystem can register a profiler, assigning it a unique name.
  That name can be used to activate the profiler or add data to it.
  The registered profiler can be composed of up to 3 functions:
    - Toggle: called when the profiler is activated/deactivated.
    - Add: called whenever data is added to the debugger
      (via `EngineDebugger::profiler_add_frame_data`)
    - Tick: called every frame (during idle), receives frame times.

- Captures: (Only relevant in remote debugger for now)
  A subsystem can register a capture, assigning it a unique name.
  When receiving a message, the remote debugger will check if it starts
  with `[prefix]:` and call the associated capture with name `prefix`.

Port MultiplayerAPI, Servers, Scripts, Visual, Performance to the new profiler system (the goal is to keep profiling code into the debugger folder, so it doesn't pollute the engine and can be completely disabled on release builds if so desired).

Port SceneDebugger and RemoteDebugger to the new capture system.
The LocalDebugger also uses the new profiler system for scripts
profiling.

Some considerations:
- **Since remote debugger now uses threads for networking**, we should review some properties like
  - max errors
  - max messages
  - max chars per seconds.
- We could reconsider some debugger features like F8 to quit the debugged game (I didn't even know we had that).
- As a next step, I'll finally integrate #33925
- **Should I rename EngineDebugger to CoreDebugger?**
- **Should I rename various `Editor*Profiler` to `EditorDebugger*Profiler`?**
- Captures are a bit specific, and maybe we could bind properties/capabilities instead.
  IIRC @Calinou and @fire suggested looking into [DAP](https://microsoft.github.io/debug-adapter-protocol/), while that seems to be a long shot, I hope this work goes into the direction of splitting different debug components apart and could potentially help in achieving such a goal (I really don't feel like working on that right now, my brain is exploding after this work :face_with_head_bandage:, sorry).

*Note*: the 1000 extra lines is the threading code, not the refactor, which in fact resulted in the same amount of code (few extra header license lines), but moved it away, whenever possible, from engine components and into the debugger folder.

### CODE REVIEW

File: core/debugger/remote_debugger.cpp [hunk 1]
```diff
@@ -0,0 +1,945 @@
...
+	struct ProfileInfoSort {
+
+		bool operator()(ScriptLanguage::ProfilingInfo *A, ScriptLanguage::ProfilingInfo *B) const {
+			return A->total_time < B->total_time;
+		}
+	};
+	Vector<ScriptLanguage::ProfilingInfo> info;
+	Vector<ScriptLanguage::ProfilingInfo *> ptrs;
+	Map<StringName, int> sig_map;
+	int max_frame_functions = 16; // TODO configurable.
```
Comment (Faless) [inline core/debugger/remote_debugger.cpp]: Aaand... I totally forgot about this, good thing I added the TODO, I'll fix this ASAP :)

**Comment by Zireael07** [comment]
Threads are YAAAY since they should massively help debugger performance in a bigger project.

**Comment by akien-mga** [comment]
CI contains about a petty style change, and unused private members: https://travis-ci.org/godotengine/godot/builds/657649282?utm_source=github_status&utm_medium=notification

**Comment by Faless** [comment]
Let's see if this fixes it, I'll try to fix my clang config.

**Comment by Faless** [comment]
Maybe now that we have threads we could also consider: #4192 (or at least a much higher max chars/max errors/max warnings defaults).

**Comment by Calinou** [comment]
@Faless I guess you should try removing the limit and see how much it can handle before stalling:

```gdscript
extends Node

func _process(delta):
    for i in 10:
        push_error(delta)
        push_warning(delta)
        print(delta)

**Comment by Faless** [comment]
@Calinou well, we still have a queue limit (or infinite loops/super verbose output might send you out of memory). I updated the PR with the removal of message-type specific limits, and a general limit of 4096 messages. Overall it seems to behave well. Now the problem is that processing messages (and errors) from the editor takes time, so the editor will go out of sync (because we have a limit in how much time it spends doing that). Few hundred errors per seconds (constantly) is fine, more will add delay, and will eventually start dropping messages (in that case, the remotedebugger will stall for few msecs, I did it on purpose, so it gets some time to resync, but we can change that).

**Comment by Faless** [comment]
I think this is ready, I've restored the old limits, raised them a bit (32 KiB per sec of text, 400 errors and 400 warnings per sec, max 2048 messages concurrently queued).
It could probably still be improved, but I feel this is already much much better then what we currently have.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**